video/out/opengl/ra_gl.c

```diff
@@ -427,40 +424,41 @@ bool ra_is_gl(struct ra *ra)
     return ra->fns == &ra_fns_gl;
 }
 
-static void gl_tex_upload(struct ra *ra, struct ra_tex *tex,
-                          const void *src, ptrdiff_t stride,
-                          struct mp_rect *rc, uint64_t flags,
-                          struct ra_buf *buf)
+static void gl_tex_upload(struct ra *ra,
+                          const struct ra_tex_upload_params *params)
 {
     GL *gl = ra_gl_get(ra);
+    struct ra_tex *tex = params->tex;
+    struct ra_buf *buf = params->buf;
     struct ra_tex_gl *tex_gl = tex->priv;
-    struct ra_buf_gl *buf_gl = NULL;
-    struct mp_rect full = {0, 0, tex->params.w, tex->params.h};
+    struct ra_buf_gl *buf_gl = buf ? buf->priv : NULL;
+    assert(tex->params.host_mutable);
 
-    if (buf) {
-        buf_gl = buf->priv;
+    const void *src = params->src;
+    if (buf)
         gl->BindBuffer(GL_PIXEL_UNPACK_BUFFER, buf_gl->buffer);
-        src = (void *)((uintptr_t)src - (uintptr_t)buf->data);
-    }
 
     gl->BindTexture(tex_gl->target, tex_gl->texture);
+    if (params->invalidate && gl->InvalidateTexImage)
+        gl->InvalidateTexImage(tex_gl->texture, 0);
 
     switch (tex->params.dimensions) {
     case 1:
-        assert(!rc);
         gl->TexImage1D(tex_gl->target, 0, tex_gl->internal_format,
                        tex->params.w, 0, tex_gl->format, tex_gl->type, src);
         break;
-    case 2:
-        if (!rc)
-            rc = &full;
-        gl_pbo_upload_tex(&tex_gl->pbo, gl, ra->use_pbo && !buf,
-                          tex_gl->target, tex_gl->format, tex_gl->type,
-                          tex->params.w, tex->params.h, src, stride,
-                          rc->x0, rc->y0, rc->x1 - rc->x0, rc->y1 - rc->y0);
+    case 2: {
+        struct mp_rect rc = {0, 0, tex->params.w, tex->params.h};
+        if (params->rc)
+            rc = *params->rc;
+        gl_upload_tex(gl, tex_gl->target, tex_gl->format, tex_gl->type,
+                      src, params->stride, rc.x0, rc.y0, rc.x1 - rc.x0,
+                      rc.y1 - rc.y0);
         break;
+    }
     case 3:
-        assert(!rc);
+        // 3D texture uploads don't really need to be efficient for us, so
+        // just use GL_UNPACK_ALIGNMENT 1 instead of worrying about it
```

Comment:
Not sure what that comment is supposed to be about. We set the alignment to 1 because we don't let the caller provide a stride, and the implied stride is the exact size of the entire line.